yarn-project/prover-node/src/bond/token-contract.ts

```diff
@@ -0,0 +1,62 @@
+import { EthAddress } from '@aztec/circuits.js';
+import { createDebugLogger } from '@aztec/foundation/log';
+import { IERC20Abi } from '@aztec/l1-artifacts';
+
+import {
+  type Chain,
+  type Client,
+  type GetContractReturnType,
+  type HttpTransport,
+  type PrivateKeyAccount,
+  type PublicActions,
+  type PublicRpcSchema,
+  type WalletActions,
+  type WalletClient,
+  type WalletRpcSchema,
+  getContract,
+} from 'viem';
+
+const MAX_ALLOWANCE = (1n << 256n) - 1n;
+const MIN_ALLOWANCE = 1n << 255n;
+
+export class TokenContract {
+  private token: GetContractReturnType<typeof IERC20Abi, WalletClient<HttpTransport, Chain, PrivateKeyAccount>>;
+  private logger = createDebugLogger('aztec:prover-node:token-contract');
+
+  constructor(
+    private readonly client: Client<
+      HttpTransport,
+      Chain,
+      PrivateKeyAccount,
+      [...WalletRpcSchema, ...PublicRpcSchema],
+      PublicActions<HttpTransport, Chain> & WalletActions<Chain, PrivateKeyAccount>
+    >,
+    address: EthAddress,
+  ) {
+    this.token = getContract({ address: address.toString(), abi: IERC20Abi, client });
+  }
+
+  /**
+   * Ensures the allowed address has near-maximum allowance, or sets it otherwise.
+   * Returns once allowance tx is mined successfully.
+   * @param allowed - Who to allow.
+   */
+  public async ensureAllowance(allowed: EthAddress) {
+    const allowance = await this.token.read.allowance([this.getSenderAddress().toString(), allowed.toString()]);
+    if (allowance < MIN_ALLOWANCE) {
+      this.logger.verbose(`Approving max allowance for ${allowed.toString()}`);
+      const hash = await this.token.write.approve([allowed.toString(), MAX_ALLOWANCE]);
```

Comment:
Couldn't we safely set this to their target balance?

Reply:
Yes, but that means that every top-up then needs 2 txs instead of 1, which means more delay